src/backend/executor/spi.c

```diff
@@ -1333,7 +1333,7 @@ SPI_cursor_open_internal(const char *name, SPIPlanPtr plan,
 	 * GPDB: Mark all queries as SPI inner queries for extension usage. But
 	 * make sure that SPI is not top level itself.
 	 */
-	if (already_under_executor_run())
+	if (!IsBackgroundWorker || _SPI_connected > 0)
```

Comment:
If I understood correctly, the `already_under_executor_run` returns an incorrect result if we are inside a function that executed inside the init plan. Is this behavior generally correct? Can you also give an example when this condition is false?

Reply:
Okay, in which case can `_SPI_connected` be 0 in this place?